bundle/src/main/java/com/adobe/acs/commons/properties/impl/PropertyAggregatorServiceImpl.java

```diff
@@ -0,0 +1,273 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import com.adobe.acs.commons.properties.PropertyAggregatorService;
+import com.day.cq.wcm.api.NameConstants;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.AttributeType;
+import org.osgi.service.metatype.annotations.Designate;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Component(service = PropertyAggregatorService.class,
+        immediate = true,
+        configurationPolicy = ConfigurationPolicy.REQUIRE)
+@Designate(ocd = PropertyAggregatorServiceImpl.Config.class)
+public class PropertyAggregatorServiceImpl implements PropertyAggregatorService {
+
+    private static final Logger log = LoggerFactory.getLogger(PropertyAggregatorServiceImpl.class);
+
+    private static final String PAGE_PROP_PREFIX = "page_properties";
+    private static final String INHERITED_PAGE_PROP_PREFIX = "inherited_page_properties";
+
+    private List<Pattern> exclusionList;
+    private Map<String, String> additionalData;
+
+    @Override
+    public Map<String, Object> getProperties(final Resource resource) {
+        Map<String, Object> map = new HashMap<>();
+
+        if (resource != null) {
+            PageManager pageManager = resource.getResourceResolver().adaptTo(PageManager.class);
+            if (pageManager != null) {
+                Page currentPage = pageManager.getContainingPage(resource);
+
+                addPagePropertiesToMap(map, currentPage, PAGE_PROP_PREFIX);
+                addInheritedPagePropertiesToMap(map, currentPage);
+                addAdditionalDataToMap(map, currentPage);
+            }
+        }
+
+        return map;
+    }
+
+    @Override
+    public Map<String, Object> getProperties(final Page page) {
+        return getProperties(page.getContentResource());
+    }
+
+    /**
+     * Iterate over the properties of a bean and add the property values (retrieved by invoking the
+     * method) to a map with the 'prefix.property' as the key.
+     * @param targetPropertyMap the map that should be updated with the properties and their values
+     * @param key the current nested name for the property (each call will append the current property name to the parent's separated by a period)
+     * @param beanValue the current property's value
+     */
+    private void addBeanToMap(final Map<String,Object> targetPropertyMap, final String key, final Object beanValue) {
```

Comment:
Why not simply have all property provider classes implement a `TemplatedDialogProperties` interface that returns the properties.  The first two implementations can be for page props and inherited page props, to abstract that out of this class which is more responsible for overall property aggregation, and would establish a more natural pattern for plugging new classes in.

Reply:
In fact those implementations could even be OSGi services themselves.  Rather than relying on string to class name configurations on the aggregation service, a dev can add new processors by simply creating instances of the TemplatedDialogProperties service, and then this aggregator service can simply inject the references.

Reply:
Even if all the objects extend an interface there would still need to be the reflection in this class to pull in the various properties. Otherwise they would have to implement the 'getProperties' method and explicitly call out what they want added. The point of the additional data configuration is that you can pass in an existing sling model and it will 'just work' simply by adding the config value. 
I can see if I can have a Sling Model also a registered OSGi service, but then it would just be using an `@Reference` of those classes to pull in and use as Sling Models instead of the string config value. If that's a desired pattern though I can do that instead!